### PR TITLE
OpenAPI-compliant REST API server with Express

### PR DESCRIPTION
**Pull request by PJColombo**

### CODE REVIEW

File: packages/api/src/context.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import { type inferAsyncReturnType } from "@trpc/server";
```
Comment (0xGabi) [inline packages/api/src/context.ts:1]: Well done!

File: packages/api/src/context.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  const innerContext = createTRPCInnerContext({ apiClient });
+
+  return {
+    ...innerContext,
+    req: opts.req,
+    res: opts.res,
+  };
+}
+
+export type TRPCContext = inferAsyncReturnType<typeof createTRPCInnerContext>;
```
Comment (0xGabi) [inline packages/api/src/context.ts:59]: This is how we get support to both API clients (express, web)?

**Review by 0xGabi — CHANGES_REQUESTED** [review]
Looking good. I left a couple of comments

File: packages/api/src/router/indexer.ts [hunk 3]
```diff
@@ -1,9 +1,9 @@
 import { z } from "zod";
 
-import { createTRPCRouter, publicProcedure } from "../trpc";
+import { createTRPCRouter, jwtAuthedProcedure, publicProcedure } from "../trpc";
 
 export const indexerRouter = createTRPCRouter({
-  getSlot: publicProcedure
+  getSlot: jwtAuthedProcedure
```
[the remark this answers is not in the document]
Comment (PJColombo) [reply packages/api/src/router/indexer.ts:6]: Mmmm, I prefer to hide any endpoint used exclusively by internal services from the public API, even if they are of type `GET`. I think it helps to maintain a clear separation of concerns and it keeps the API well-organized :thinking:

File: packages/api/src/context.ts (hunk 2 above)
Comment (PJColombo) [reply packages/api/src/context.ts:59]: `inferAsyncReturnType` is just a little helper provided by trpc to get the context type itself.

Right now, the way we support both express and nextjs consumers is by doing [this](https://github.com/Blobscan/blobscan/blob/334c77ac65516b84854ee3884684107d3da075a4/packages/api/src/context.ts#L14).

As our procedures logic does not rely on any specific features of Express or Next.js, this approach has not caused any issues. However, we may need to reevaluate and consider adopting a more framework-agnostic strategy in the future.

File: packages/api/src/trpc.ts [hunk 4]
```diff
@@ -132,13 +112,7 @@ export const publicProcedure = t.procedure;
...
- * this. It verifies the session is valid and guarantees ctx.session.user is not
- * null
- *
- * @see https://trpc.io/docs/procedures
- */
-// export const protectedProcedure = t.procedure.use(enforceUserIsAuthed);
+export const createTRPCRouter = t.router;
+
+export const publicProcedure = t.procedure;
+export const jwtAuthedProcedure = t.procedure.use(isJWTAuthed);
```
[the remark this answers is not in the document]
Comment (PJColombo) [reply packages/api/src/trpc.ts:118]: Forgot to handle the jwt verification errors. Just pushed a commit with the fix.

The indexer sets up the expiration date of the JWT tokens being sent to the REST API. The tokens are refreshed every certain amount of time (5 or more minutes for example). Both services share the same secret key which they use to encode and decode the JWT tokens.
The rationale behind this is that we have a simple architecture where only two services need to communicate with each other, so I don't see the need of having an auth server provider such as Auth0 to handle this logic. We also wanted to handle the auth ourselves.

The TTL checks are handled in the jwt's [`verify`](https://github.com/Blobscan/blobscan/blob/334c77ac65516b84854ee3884684107d3da075a4/packages/api/src/context.ts#L35) function. It throws an error if the token has expired

[the remark this answers is not in the document]
Comment (0xGabi) [reply packages/api/src/trpc.ts:118]: >The rationale behind this is that we have a simple architecture where only two services need to communicate with each other, so I don't see the need of having an auth server provider such as Auth0 to handle this logic. We also wanted to handle the auth ourselves.

Make sense. Thanks for clarifying the flow.

**Pull request merged**